packages/expo-updates/android/src/main/java/expo/modules/updates/loader/FileDownloader.kt

```diff
@@ -70,38 +81,91 @@ class FileDownloader(
   /**
    * Constructor for tests
    */
-  constructor(filesDirectory: File, easClientID: String, configuration: UpdatesConfiguration, logger: UpdatesLogger, client: OkHttpClient) : this(filesDirectory, easClientID, configuration, logger) {
+  constructor(filesDirectory: File, easClientID: String, configuration: UpdatesConfiguration, logger: UpdatesLogger, client: OkHttpClient) : this(filesDirectory, easClientID, configuration, logger, null) {
     this.client = client
   }
 
   data class FileDownloadResult(val file: File, val hash: ByteArray)
   data class AssetDownloadResult(val assetEntity: AssetEntity, val isNew: Boolean)
 
   private suspend fun downloadAssetAndVerifyHashAndWriteToPath(
+    asset: AssetEntity,
     request: Request,
     expectedBase64URLEncodedSHA256Hash: String?,
     destination: File,
-    progressListener: FileDownloadProgressListener? = null
+    updatesDirectory: File,
+    progressListener: FileDownloadProgressListener? = null,
+    allowPatch: Boolean = true
   ): FileDownloadResult {
     try {
       val response = downloadData(request, progressListener)
 
-      if (!response.isSuccessful) {
-        val message = "Asset download request not successful"
-        val cause = IOException(response.body?.string() ?: "Unknown error")
-        logger.error(message, cause, UpdatesErrorCode.AssetsFailedToLoad)
-        throw IOException(message, cause)
-      }
+      response.use { resp ->
+        if (!resp.isSuccessful) {
+          val message = "Asset download request not successful"
+          val cause = IOException(resp.body?.string() ?: "Unknown error")
+          logger.error(message, cause, UpdatesErrorCode.AssetsFailedToLoad)
+          throw IOException(message, cause)
+        }
 
-      try {
-        response.body!!.byteStream().use { inputStream ->
-          val hash = UpdatesUtils.verifySHA256AndWriteToFile(inputStream, destination, expectedBase64URLEncodedSHA256Hash)
-          return FileDownloadResult(destination, hash)
+        val responseBody = resp.body
+          ?: throw IOException("Asset download response from ${request.url} had no body")
+
+        val isPatch = isPatchResponse(responseBody)
+        val hash = if (allowPatch && isPatch) {
+          try {
+            applyPatchAndVerify(
+              asset,
+              request,
+              responseBody,
+              destination,
+              updatesDirectory,
+              expectedBase64URLEncodedSHA256Hash
+            )
+          } catch (_: Exception) {
```

Comment:
Not sure if this will mask other errors and accidentally trigger refetches for things that should've just thrown

Reply:
Took the approach of always being defensive with diffing. If anything goes wrong, we fall back to the normal flow